R/exploratoryfactoranalysis.R

```diff
@@ -413,6 +491,70 @@ ExploratoryFactorAnalysis <- function(jaspResults, dataset, options, ...) {
 
 }
 
+# Modification here:
+# Creates a table that outputs the details of the parallel analysis.
+# Particularly useful for reporting the rationale of how many factors were
+# retained - especially within psychometric papers.
+# We are also going to be adding an asterisk aside any factor whose real data eigenvalue is above
+# the 95th percentile simulated data eigenvalue
+.efaPATable <- function(modelContainer, dataset, options, ready) {
+  if (!options[["incl_PAtable"]] || !is.null(modelContainer[["patab"]])) return()
+
+  if (options[["basedOn"]] == "mixed") {
+    polytetracor <- psych::mixedCor(dataset)
+    pa <- try(psych::fa.parallel(polytetracor$rho,
+                                 plot = FALSE,
+                                 fa = options$parallelMethod,
+                                 n.obs = nrow(dataset)))
+  }
+  else {
+    pa <- try(psych::fa.parallel(dataset, plot = FALSE, fa = options$parallelMethod))
+  }
+
+  patab <- createJaspTable(gettext("Parallel Analysis"))
+  patab$dependOn("incl_PAtable")
+  patab$addColumnInfo(name = "col", title = "", type = "string")
+  patab$addColumnInfo(name = "val1", title = gettext("Real Data Eigenvalue"), type = "number", format = "dp:3")
+  patab$addColumnInfo(name = "val2", title = gettext("Mean of Resampled Eigenvalues"), type = "number", format = "dp:3")
+  patab$position <- 5
+  modelContainer[["patab"]] <- patab
+
+  if (!ready || modelContainer$getError()) return()
+
+  if (options$parallelMethod == "pc") {
+    RealDataEigen <- pa$pc.values
+    ResampledEigen <- pa$pc.sim
+  } else { # parallelmethod is fa
+    RealDataEigen <- pa$fa.values
+    ResampledEigen <- pa$fa.sim
+  }
+
+  efaResult <- modelContainer[["model"]][["object"]]
+  PAs <- zapsmall(as.matrix(data.frame(RealDataEigen,ResampledEigen),fix.empty.names = F))
+  forasterisk <- data.frame(applyforasterisk=RealDataEigen-ResampledEigen)
+  dims <- nrow(PAs)
+
+  firstcol <- data.frame()
+  for (i in 1:dims) {
+    if (forasterisk$applyforasterisk[i] > 0) {
+      firstcol <- rbind(firstcol,paste("Factor", " ", i,"*", sep=""))
+    }
+    else {
+      firstcol <- rbind(firstcol,paste("Factor", i))
+    }
+  }
+  patab[["col"]] <- c(firstcol$X.Factor.1..)
```

Comment:
```suggestion
   if (options$parallelMethod == "pc") {
    eigTitle <- gettext("Real data component eigenvalues")
    rowsName <- gettext("Component")
    RealDataEigen <- pa$pc.values
    ResampledEigen <- pa$pc.sim
    footnote <- gettext("'*' = Component should be retained.\nResults from PC-based parallel analysis")
  } else { # parallelmethod is FA
    eigTitle <- gettext("Real data factor eigenvalues")
    rowsName <- gettext("Factor")
    RealDataEigen <- pa$fa.values
    ResampledEigen <- pa$fa.sim
    footnote <- gettext("'*' = Factor should be retained.\nResults from FA-based parallel analysis")
  }

  patab <- createJaspTable(gettext("Parallel Analysis"))
  patab$dependOn("incl_PAtable")
  patab$addColumnInfo(name = "col", title = "", type = "string")

  patab$addColumnInfo(name = "val1", title = eigTitle, type = "number", format = "dp:3")
  patab$addColumnInfo(name = "val2", title = gettext("Simulated data mean eigenvalues"), type = "number", format = "dp:3")
  patab$position <- 5
  modelContainer[["patab"]] <- patab

  if (!ready || modelContainer$getError()) return()

  efaResult <- modelContainer[["model"]][["object"]]
  PAs <- zapsmall(as.matrix(data.frame(RealDataEigen, ResampledEigen), fix.empty.names = FALSE))
  forasterisk <- data.frame(applyforasterisk = RealDataEigen - ResampledEigen)
  dims <- nrow(PAs)

  firstcol <- data.frame()
  for (i in 1:dims) {
    if (forasterisk$applyforasterisk[i] > 0) {
      firstcol <- rbind(firstcol, paste(rowsName, " ", i,"*", sep = ""))
    } else {
      firstcol <- rbind(firstcol, paste(rowsName, i))
    }
  }

  patab[["col"]] <- firstcol[[1]]
```

Reply:
This looks like a lot of changes but it is actually not. I just changed the order of things a bit and added the distinction between components and factors for the parallel analysis. I am actually not 100% certain if the table should definitively say "components" (when "based on pc" is selected) instead of factors when the original goal is to know the number of factors to estimate. Then again, when the parallel analysis is based on components this would be the theoretically correct naming. Having said that, I dont do EFA often so what would people expect here?

Reply:
> I am actually not 100% certain if the table should definitively say "components" (when "based on pc" is selected) instead of factors when the original goal is to know the number of factors to estimate. Then again, when the parallel analysis is based on components this would be the theoretically correct naming. Having said that, I dont do EFA often so what would people expect here?

You're technically 100% correct - when using PC-based parallel analysis, the estimated eigenvalues correspond to components rather than factors; however, since it is also typically used as a valid method of estimating the number of factors within EFA, it sounds like maybe the word "factor" or something alike would be suitable here - I believe either way people would understand. Maybe use the word "dimension"? 

I'm basing this table on the software FACTOR which produces a similar output - it uses the word "variable" which in my opinion is actually somewhat less precise than "factor/component/dimension".

Reply:
That is a tricky one. If we name it "factor" throughout, it is technically incorrect, but what people are used to. If we distinguish between "components" and "factors" we risk confusing people... Any preference @AlexanderLyNL ?

Reply:
I'll keep it as "factor" and will add a brief remark in the help file. Basically, this is an analysis where we see how many factors could explain more variance than randomized data. On the one hand, we have eigenvalues corresponding to our data and then eigenvalues extracted from the mean or top 95th percentile of "bootstrapped" randomized data. The eigenvalues can come from a factor analysis or a principal components analysis, but both methods are used to assess the number of factors within EFA (hence the uncertainty while naming the footnote - "factor/component"). This table in particular shows the eigenvalues of both factorized "datasets" (real vs. randomized); the real-data factor/component eigenvalues  that surpass the randomized-data factor/component eigenvalues are deemed advisable for extraction.